### PR TITLE
[WIP] Run commands via issue comments

### PR DESCRIPTION
**Pull request by rudehn**
Start working towards #192 by retrieving issue comments

So what's going on here? Good question!

This provides the ability to run commands to chaosbot through issue comments. Current functionality enables closing and (maybe?) opening of issues, however this can be expanded to be as flexible as needed.

This could definitely use some review and testing from other people.

### CODE REVIEW

**Comment by hongaar** [comment]
Awesome! Was just about to start working on this, too ;)

**Comment by hongaar** [comment]
@rudehn like where this is going! Maybe add [WIP] in your title to hold off voting window for now... As for a list of commands:

- `/vote close` closes issue when no nay reactions on this comment are added within voting window
- `/vote reopen` reopens issue when see above ^^^
- `/vote label=<LABEL_TEXT>` adds label when ^^^
- `/vote remove-label=<LABEL_TEXT>` removes label when ^^^
- `/vote assign=<USER>` assigns to user when ^^^
- `/vote unassign=<USER>` unassigns from user when ^^^

Multiple commands could be added in a single comment, nay votes would then apply to all commands at once.

Maybe only implement closing first as a POC?

**Comment by rudehn** [comment]
The only issue is this doesn't support getting reactions on the comments. Haven't figured that one out yet

**Comment by rudehn** [comment]
Ok this could definitely use a review.

**Comment by PlasmaPower** [comment]
Instead of no -1s, how about a positive total? (counting the author as an implicit +1)

**Comment by rudehn** [comment]
Sure that's definitely a possibility. This method is just implementing  #192. If the community comes up with a better way, then we will go with that.

**Comment by hongaar** [comment]
Hmm.. why not? Can't test the code right now, its way too sunny outside, but looking good! One question: how does it track which commands did already pass before? If say a reopen command was passed, and the author closes the issue later on, won't it automatically reopen it again?

**Comment by PlasmaPower** [comment]
It looks like you're right: a command will be continually reapplied until down voted. That could cause problems.

**Comment by rudehn** [comment]
Yeah it definitely will haha. This is a first pass for POC with respect to closing issues and getting a general framework in place for submitting commands.

I'll look into saving the previously ran commands into a json file. Would be nice if we had mysql support up..

**Comment by PlasmaPower** [comment]
We do have MySQL IIRC, just no interface to it.

**Comment by reddraggone9** [comment]
You could also have @chaosbot add a reaction to record that it ran/didn't run the command.

**Comment by rudehn** [comment]
@reddraggone9 that's true. 

I was going to suggest we keep a timestamp that indicates where we stopped at last time we ran commands. We could sort the issue comments by "created_at" date. And wait for them to go through the voting window. Then we only retrieve new issue comments after our previous timestamp we left off at.

This would probably interfere with updating commands and resetting the voting window etc...

A safer solution I think would be to just save the comment id and never run the command from that comment id again.

**Comment by hongaar** [comment]
@rudehn saving the comment IDs as a means of recording them as 'already processed, ignore from now on' sounds good to me. Would be great if this list could be persisted in the repo itself? (i.e. chaosbot updates a text file and pushes that up to the GH repo?)

**Comment by amoffat** [comment]
Very cool stuff @rudehn!  If you're using the docker dev env to test your stuff, let me know how that works out for you, so it can be improved

File: cron/poll_read_issue_comments.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        elif(sub_command == "reopen"):
+            if(can_run_vote_command(votes, comment_id)):
+                gh.issues.open_issue(api, settings.URN, issue_id)
+        else:
+            # Other commands have an = in them
+            sub_command = sub_command.split("=")
+            if(len(sub_command > 1)):
+                args = sub_command[1:]
+                sub_command = sub_command[0]
+                if(sub_command == "label"):
```
Comment (xyproto) [inline cron/poll_read_issue_comments.py]: A bit shorter:

```python
if sub_command[0] not in ["label", "remove-label", "assign", "unassign"]:
    log_warning = True
```

Comment (rudehn) [reply cron/poll_read_issue_comments.py]: it is, but you still have to check which command you need to run. These commands would be implemented sometime in the future

File: cron/poll_read_issue_comments.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    votes = {}
+    for voter, vote in gh.voting.get_comment_reaction_votes(api, urn, comment_id):
+        votes[voter] = vote
+    return votes
+
+def handle_vote_command(api, command, issue_id, comment_id, votes):
+    orig_command = command[:]
+    # Check for correct command syntax, ie, subcommands
+    log_warning = False
+    if(len(command)):
```
Comment (aslafy-z) [inline cron/poll_read_issue_comments.py]: Please, write pythonic code (remove the parentheses around the ifs) !

File: cron/poll_read_issue_comments.py (hunk 1 above)
Comment (xyproto) [reply cron/poll_read_issue_comments.py]: I think it's better to implement things as they are needed. Scaffolding for the future can easily create unmanagable code bases over time. Someone will have to remember to actually do the implementation sometime in the future, and this does not always happen.

File: cron/poll_read_issue_comments.py (hunk 2 above)
Comment (rudehn) [reply cron/poll_read_issue_comments.py]: Ahh my bad!! I totally agree.

**Comment by rudehn** [comment]
Looking to finish this up soon. Been a busy weekend

**Comment by rudehn** [comment]
Alright its up! Voting commands for closing and reopening issues is working.

Todo

- [ ] Add chaos response to issue command similar to PR status

**Comment by phil-r** [comment]
@rudehn Something is wrong with a PR 😕

**Comment by rudehn** [comment]
Certainly looks like it. I'll see if I can figure out what happened

File: cron/poll_read_issue_comments.py [hunk 3]
```diff
@@ -67,29 +86,29 @@ def get_command_votes(api, urn, comment_id):
...
+            gh.issues.close_issue(api, settings.URN, issue_id)
+            update_command_ran(comment_id)
+        elif sub_command == "reopen":
+            gh.issues.open_issue(api, settings.URN, issue_id)
+            update_command_ran(comment_id)
         else:
             # Other commands have an = in them
             sub_command = sub_command.split("=")
-            if(len(sub_command > 1)):
+            if len(sub_command > 1):
```
Comment (aslafy-z) [inline cron/poll_read_issue_comments.py]: Should be `len(sub_command) > 1:` right ?

File: cron/poll_read_issue_comments.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+        if sub_command == "close":
+            gh.issues.close_issue(api, settings.URN, issue_id)
+            update_command_ran(comment_id)
+        elif sub_command == "reopen":
+            gh.issues.open_issue(api, settings.URN, issue_id)
+            update_command_ran(comment_id)
+        else:
+            # Other commands have an = in them
+            sub_command = sub_command.split("=")
+            if len(sub_command > 1):
```
Comment (aslafy-z) [inline cron/poll_read_issue_comments.py]: Should be `len(sub_command)  > 1:` right ?

**Review by phil-r — CHANGES_REQUESTED** [review]
@rudehn I'm forced to downvote, until the mess(110 commits and 44 files changed) is fixed 👎

**Comment by phil-r** [comment]
@rudehn maybe this will help: #339

**Comment by aslafy-z** [comment]
Please switch back your PR to WIP until it's clean @rudehn

**Comment by rudehn** [comment]
I'm not sure how to fix it. I don't really know what happened

**Pull request closed without merging**